maven_iuvs/echelle.py

```diff
@@ -0,0 +1,886 @@
+               metadata for all the dark observation files
+    """
+    dark_idx = [fidx for fidx in ech_l1a_idx if (('orbit' in fidx['name']) and ech_isdark(fidx))]
+    dark_idx = sorted(dark_idx, key=lambda i:i['datetime'])
+    
+    return dark_idx
+
+# Count rates ----------------------------------------------------------
+
+
+def get_avg_pixel_count_rate(hdul, spapixrange, spepixrange, return_npix=True):
+    """
+    Determines the DN count rate (DN/pixel/s) over the given spatial x spectral ranges
+    for the data contained in hdul.
+
+    Parameters
+    ----------
+    hdul : astropy FITS HDUList object
+           HDU list for a given observation
+    spapixrange : array
+                  Pixels which define the slit start and end.
+    spepixrange : array
+                  Pixels which define the spectral window over which
+                  we can expect to find the emisssion.
+    return_npix : Whether to return the total number of pixels in the 
+                  spatial x spectral window defined, 
+                  i.e. product of nspapix * nspepix.
+
+    Returns
+    -------
+    countrate : DN per pixel per second
+    npix : number of pixels contained in the spatial x spectral window.
+
+    """
+    binning = get_binning_scheme(hdul)
+    n_int = get_n_int(hdul)
+    
+    # Retrieve indices at which to index the binned data 
+    spalo, spahi = spapixrange
+    spabinlo, spabinhi, nspapix = pix_to_bin(hdul,
+                                             spalo, spahi, 'SPA')
+    spelo, spehi = spepixrange
+    spebinlo, spebinhi, nspepix = pix_to_bin(hdul, 
+                                             spelo, spehi, 'SPE')
+
+    npix = nspapix*nspepix
+
+    if binning['nspa'] == 0 or binning['nspe'] == 0:
+        # data is bad and contains no frames
+        countsperpix = np.nan
+    elif n_int == 1:
+        # single integration
+        countsperpix = np.sum(hdul['Primary'].data[spabinlo:spabinhi, spebinlo:spebinhi])/npix
+    else: # n_int > 1
+        countsperpix = np.sum(hdul['Primary'].data[:, spabinlo:spabinhi, spebinlo:spebinhi], axis=(1,2))/npix    
+        
+    countrate = np.atleast_1d(countsperpix)/hdul['Primary'].header['INT_TIME']
+    
+    if return_npix:
+        return countrate, npix
+    
+    return countrate
+
+
+def get_countrate_diagnostics(hdul):
+    """
+    Produces the count rate in DN/pix/s in the H and D emissions, as 
+    well as in the background near the emissions.
+
+    Parameters
+    ----------
+    hdul : astropy FITS HDUList object
+           HDU list for a given observation
+
+    Returns
+    -------
+    dictionary
+           containing the count rates and number of pixels included 
+           in areas on the detector covering:
+           (1) H Ly alpha emission;
+           (2) H Ly alpha background;
+           (3) D Ly alpha emission;
+           (4) D Ly alpha background.
+
+
+    """
+    Hlya_spapixrange = np.array([ech_slit_start, ech_slit_end])
```

Comment:
need to add an exception to abort if file is not ech, since the pix ranges make no sense for FUV or MUV data

Reply:
is it not enough to already be in the echelle module? Wouldn't we thus want to apply similar error handling in other functions?

Reply:
Yeah, probably fine to just trust the user in this case.